### PR TITLE
Python: Handle `formattedValue`

### PR DESCRIPTION
**Pull request by lshala**
This PR adds the handling for `formattedValue`'s.

### CODE REVIEW

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/ExpressionHandler.kt [hunk 1]
```diff
@@ -182,45 +182,83 @@ class ExpressionHandler(frontend: PythonLanguageFrontend) :
         return assignExpression
     }
 
+    /**
+     * Translates a Python
+     * [`FormattedValue`](https://docs.python.org/3/library/ast.html#ast.FormattedValue) into a
+     * [Expression].
+     *
+     * We are handling the format handling, following [PEP 3101](https://peps.python.org/pep-3101).
```
Comment (maximiliankaul) [inline cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/ExpressionHandler.kt:190]: Is this the correct PEP? I don't find any mentions of `ascii` in the document.

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/ExpressionHandler.kt [hunk 2]
```diff
@@ -182,45 +182,83 @@ class ExpressionHandler(frontend: PythonLanguageFrontend) :
...
+                else ->
+                    newProblemExpression(
+                        "Cannot handle formatted value with conversion ${node.conversion} yet",
+                        rawNode = node
+                    )
             }
-            else ->
-                newProblemExpression(
-                    "Cannot handle formatted value with conversion ${node.conversion} yet",
+        if (formatSpec != null) {
```
Comment (maximiliankaul) [inline cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/ExpressionHandler.kt:243]: Probably good to add some error handling. The AST doc says this is a `JoinedStr` and we treat it like that. However, it'd be good to catch some errors here for the `format` call to make sense!?

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/ExpressionHandler.kt [hunk 3]
```diff
@@ -182,45 +182,83 @@ class ExpressionHandler(frontend: PythonLanguageFrontend) :
         return assignExpression
     }
 
+    /**
```
Comment (maximiliankaul) [inline cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/ExpressionHandler.kt:185]: Add some short example how this function models Python code -> CPG nodes.

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/ExpressionHandler.kt [hunk 4]
```diff
@@ -182,45 +181,84 @@ class ExpressionHandler(frontend: PythonLanguageFrontend) :
...
+                    addArgument(conversion)
+                    addArgument(formatSpec)
+                }
         }
+        return conversion
     }
 
+    /**
+     * Translates a Python [`JoinedStr`](https://docs.python.org/3/library/ast.html#ast.JoinedStr)
+     * into a [Expression].
```
Comment (maximiliankaul) [inline cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/ExpressionHandler.kt:260]:
```suggestion
     * into an [Expression].
```

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/ExpressionHandler.kt [hunk 5]
```diff
@@ -182,45 +181,84 @@ class ExpressionHandler(frontend: PythonLanguageFrontend) :
...
-                newProblemExpression(
-                    "Cannot handle formatted value with conversion ${node.conversion} yet",
+        if (formatSpec != null) {
+            return newCallExpression(
+                    callee = newReference(name = "format", rawNode = node),
+                    fqn = "format",
                     rawNode = node
                 )
+                .implicit()
+                .apply {
```
Comment (maximiliankaul) [inline cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/ExpressionHandler.kt:250]: Is this order correct?

File: cpg-language-python/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/python/PythonFrontendTest.kt [hunk 6]
```diff
@@ -1376,6 +1377,93 @@ class PythonFrontendTest : BaseTest() {
         assertEquals(2L, fStmtRhsThird.value)
     }
 
+    @Test
+    fun testFormattedValues() {
```
Comment (maximiliankaul) [inline cpg-language-python/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/python/PythonFrontendTest.kt:1381]: Move to ExpressionHandler test?

File: cpg-language-python/src/test/resources/python/datatypes.py [hunk 7]
```diff
@@ -7,4 +7,10 @@
     "e": "f"
 }
 e = f'Values of a: {a} and b: {b!s}'
-f = a[1:3:2]
\ No newline at end of file
+f = a[1:3:2]
```
Comment (maximiliankaul) [inline cpg-language-python/src/test/resources/python/datatypes.py:10]: Please have a separate test code file specific for this test. Reusing this datatypes.py should be avoided.

File: cpg-language-python/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/python/PythonFrontendTest.kt [hunk 8]
```diff
@@ -1376,6 +1377,93 @@ class PythonFrontendTest : BaseTest() {
...
+                true
+            ) {
+                it.registerLanguage<PythonLanguage>()
+            }
+        assertNotNull(tu)
+        val namespace = tu.namespaces.singleOrNull()
+        assertNotNull(namespace)
+
+        // Test for g = f'Number: {42:.2f}'
+        val gStmt = namespace.statements[6]
```
Comment (maximiliankaul) [inline cpg-language-python/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/python/PythonFrontendTest.kt:1396]: I'm not too happy with the index access. However, I cannot provide a good alternative. The best thing I can come up with is:
```python
val gStmt = tu.refs["g"]?.astParent
```
Do you think this is better? Unfortunately, we lack any means to get the declaration here.

Comment (maximiliankaul) [reply cpg-language-python/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/python/PythonFrontendTest.kt:1396]: Or better `tu.variables["g"]?.astParent`

File: cpg-language-python/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/python/PythonFrontendTest.kt [hunk 9]
```diff
@@ -1376,6 +1377,93 @@ class PythonFrontendTest : BaseTest() {
...
+        assertIs<AssignExpression>(gStmt)
+        val gStmtRhs = gStmt.rhs.singleOrNull()
+        assertIs<BinaryOperator>(gStmtRhs)
+        val gFormatCall = gStmtRhs.rhs
+        assertIs<CallExpression>(gFormatCall)
+        assertEquals("format", gFormatCall.name.localName)
+        val gArguments = gFormatCall.arguments
+        assertEquals(2, gArguments.size)
+        assertIs<Literal<*>>(gArguments[0])
+        assertEquals(42.toLong(), gArguments[0].value.value)
```
Comment (maximiliankaul) [inline cpg-language-python/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/python/PythonFrontendTest.kt:1406]:
```suggestion
                assertLiteralValue(42.toLong(), gArguments[0])
```
Please also change the other occurrences.

**Review by maximiliankaul — COMMENTED** [review]
Overall, this looks good to me. There are some confusions / bugs that have to be resolved before approving, though.

File: cpg-language-python/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/python/PythonFrontendTest.kt [hunk 10]
```diff
@@ -1376,6 +1377,93 @@ class PythonFrontendTest : BaseTest() {
...
+        assertIs<Literal<*>>(gArguments[1])
+        assertEquals(".2f", gArguments[1].value.value)
```
Comment (maximiliankaul) [inline cpg-language-python/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/python/PythonFrontendTest.kt:1408]: I don't think the formatting specifier (or however this is called) belongs to the format call as an argument.

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/ExpressionHandler.kt (hunk 2 above)
Comment (maximiliankaul) [reply cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/ExpressionHandler.kt:243]: Not the correct place (above for handle).

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/ExpressionHandler.kt (hunk 5 above)
Comment (maximiliankaul) [reply cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/ExpressionHandler.kt:250]: Yes, it is.

File: cpg-language-python/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/python/PythonFrontendTest.kt [hunk 11]
```diff
@@ -1376,6 +1377,93 @@ class PythonFrontendTest : BaseTest() {
...
+        val jStmtRhs = jStmt.rhs.singleOrNull()
+        assertIs<BinaryOperator>(jStmtRhs)
+        val jConversionCall = jStmtRhs.rhs
+        assertIs<CallExpression>(jConversionCall)
+        assertEquals("ascii", jConversionCall.name.localName)
+        val jArg = jConversionCall.arguments.singleOrNull()
+        assertNotNull(jArg)
+        assertEquals("d", jArg.name.localName)
+
+        // Test for k = f'Combined: {b!s:10}'
```
[the remark this answers is not in the document]
Comment (maximiliankaul) [reply cpg-language-python/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/python/PythonFrontendTest.kt:1449]: I misunderstood the CPG translation.

**Review by maximiliankaul — APPROVED** [review]
Looks good to merge once the conversations have been resolved. Nice PR :+1:

File: cpg-language-python/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/python/PythonFrontendTest.kt (hunk 11 above)
Comment (maximiliankaul) [inline cpg-language-python/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/python/PythonFrontendTest.kt:1449]:
```suggestion
        // Test for k = f'Combined: {b!s:10}'
        // This is translated to `'Combined: ' +  format(str(b), "10")`
```

File: cpg-language-python/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/python/PythonFrontendTest.kt (hunk 6 above)
Comment (lshala) [reply cpg-language-python/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/python/PythonFrontendTest.kt]: Done.

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/ExpressionHandler.kt (hunk 3 above)
Comment (lshala) [reply cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/ExpressionHandler.kt:171]: Done. Is this what you had in mind?

Comment (maximiliankaul) [reply cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/ExpressionHandler.kt:171]: Looks good to me. I would have added:

The code `f"{value:.2f}"` is first transformed to the equivalent expression `format(value, ".2f")` and then modeled as:

This has the `format(value, ".2f")` translation in the example, which I think is easier to read. But I'm happy with the current doc, too :)

**Pull request merged**